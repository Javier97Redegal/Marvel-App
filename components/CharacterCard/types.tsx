type CharacterCardProps = {
    character: {
        id: number
        name: string | undefined
        thumbnail: {
            path: string
            extension: string
        }
    }
}