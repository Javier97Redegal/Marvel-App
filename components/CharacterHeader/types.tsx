export interface CharacterHeaderProps {
    id: number
}