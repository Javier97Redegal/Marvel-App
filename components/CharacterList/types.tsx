export interface CharacterListProps {
    mode: string
    query: string
}