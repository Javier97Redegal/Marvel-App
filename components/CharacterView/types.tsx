export interface CharacterViewProps {
    id: string
}