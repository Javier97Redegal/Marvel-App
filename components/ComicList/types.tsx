export interface ComicListProps {
    id: number
}