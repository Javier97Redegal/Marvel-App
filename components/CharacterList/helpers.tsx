export const getResultsMessage = (number: number): string => {
    return `Result${number === 1 ? '' : 's'}`
}