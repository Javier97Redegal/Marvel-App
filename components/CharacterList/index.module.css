.header {
    max-width: 1980px;
    padding: 3rem 2rem 0 2rem;
    margin: 0 auto;
    text-transform: uppercase;
    font-size: 2rem;
    font-weight: 900;
}

.totals {
    display: block;
    text-transform: uppercase;
    font-size: 1.225rem;
    max-width: 1980px;
    margin: 2rem 2rem 0 2rem;
}

.list {
    display: grid;
    grid-template-columns: repeat(7, minmax(0 ,1fr));
    gap: 1.5rem;
    padding: 2rem;
    list-style: none;
    max-width: 1980px;
    margin: 0 auto;
}