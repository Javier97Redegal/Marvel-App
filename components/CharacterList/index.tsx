import CharacterCard from '@/components/CharacterCard'
import { CharacterListProps } from './types'

const CharacterList = ({ list }: CharacterListProps) => <>
    <ul>
        {list.map((character, key) => (
            <li key={key}>
                <CharacterCard {...character} />
            </li>
        ))}
    </ul>
</>

export default CharacterList