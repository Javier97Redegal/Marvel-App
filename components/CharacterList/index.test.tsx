import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { useQuery } from 'react-query'
import CharacterList from '@/components/CharacterList'
import { CharacterListProps } from '@/components/CharacterList/types'
import { useApp } from '@/context/AppContext'
import { TYPES_OF_CHARACTER_LIST } from '@/constants'

jest.mock('react-query', () => ({
    useQuery: jest.fn()
}))

jest.mock('@/context/AppContext', () => ({
    useApp: jest.fn()
}))

jest.mock('@/utils/MarvelAPI', () => ({
    getCharacters: jest.fn()
}))

jest.mock('@/components/CharacterCard', () => jest.fn(({ id }: { id: number }) => <div>CharacterCard Mock {id}</div>))
jest.mock('@/components/LoadingPanel', () => jest.fn(({ type }) => <div>LoadingPanel Mock: {type}</div>))

describe('CharacterList test', () => {
    const mockProps: CharacterListProps = {
        mode: TYPES_OF_CHARACTER_LIST.TOTAL,
        query: ''
    }

    const charactersData = {
        data: {
            results: [
                { id: '1', name: 'Spider-Man', thumbnail: { path: 'path/to/spiderman', extension: 'jpg' } },
                { id: '2', name: 'Iron Man', thumbnail: { path: 'path/to/ironman', extension: 'jpg' } }
            ]
        }
    }

    const mockUseApp = {
        favorites: [{ id: '1', name: 'Spider-Man', thumbnail: { path: 'path/to/spiderman', extension: 'jpg' } }],
    }

    beforeEach(() => {
        (useApp as jest.Mock).mockReturnValue(mockUseApp)
    })

    it('renders loading panel when loading', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'loading' })

        const { asFragment } = render(<CharacterList {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: loading')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders loading panel when error', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: null, status: 'error' })

        const { asFragment } = render(<CharacterList {...mockProps} />)

        expect(screen.getByText('LoadingPanel Mock: error')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders character cards when data is loaded', () => {
        (useQuery as jest.Mock).mockReturnValue({ data: charactersData, status: 'success' })

        const { asFragment } = render(<CharacterList {...mockProps} />)

        expect(screen.getByText('2 Results')).toBeInTheDocument()
        expect(screen.getByText('CharacterCard Mock 1')).toBeInTheDocument()
        expect(screen.getByText('CharacterCard Mock 2')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders favorites when mode is FAVORITES', () => {
        const favoriteProps = { ...mockProps, mode: TYPES_OF_CHARACTER_LIST.FAVORITES }

        render(<CharacterList {...favoriteProps} />)

        expect(screen.getByText('1 Result')).toBeInTheDocument()
        expect(screen.getByText('Favorites')).toBeInTheDocument()
        expect(screen.getByText('CharacterCard Mock 1')).toBeInTheDocument()
    })

    it('filters favorites based on query', () => {
        const favoriteProps = { ...mockProps, mode: TYPES_OF_CHARACTER_LIST.FAVORITES, query: 'man' }

        render(<CharacterList {...favoriteProps} />)

        expect(screen.getByText('1 Result')).toBeInTheDocument()
        expect(screen.getByText('CharacterCard Mock 1')).toBeInTheDocument()
    })

    it('displays no results when favorites do not match query', () => {
        const favoriteProps = { ...mockProps, mode: TYPES_OF_CHARACTER_LIST.FAVORITES, query: 'batman' }

        render(<CharacterList {...favoriteProps} />)

        expect(screen.getByText('0 Results')).toBeInTheDocument()
        expect(screen.getByText('LoadingPanel Mock: success')).toBeInTheDocument()
    })
})