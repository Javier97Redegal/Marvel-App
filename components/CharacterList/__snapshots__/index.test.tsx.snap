// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList test renders character cards when data is loaded 1`] = `
<DocumentFragment>
  <div
    class="totals"
  >
    2 Results
  </div>
  <ul
    class="list"
  >
    <li>
      <div>
        CharacterCard Mock 1
      </div>
    </li>
    <li>
      <div>
        CharacterCard Mock 2
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`CharacterList test renders loading panel when error 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: error
  </div>
</DocumentFragment>
`;

exports[`CharacterList test renders loading panel when loading 1`] = `
<DocumentFragment>
  <div>
    LoadingPanel Mock: loading
  </div>
</DocumentFragment>
`;
